libafl/src/events/llmp/hooks/centralized.rs

```diff
@@ -0,0 +1,117 @@
+use std::{fmt::Debug, marker::PhantomData};
+
+#[cfg(feature = "llmp_compression")]
+use libafl_bolts::{compress::GzipCompressor, llmp::LLMP_FLAG_COMPRESSED};
+use libafl_bolts::{
+    llmp::{Flags, LlmpHook, LlmpMsgHookResult, Tag},
+    shmem::ShMemProvider,
+    ClientId, Error,
+};
+
+#[cfg(feature = "llmp_compression")]
+use crate::events::COMPRESS_THRESHOLD;
+use crate::{
+    events::{BrokerEventResult, Event, _LLMP_TAG_TO_MAIN},
+    inputs::Input,
+};
+
+/// An LLMP-backed event manager for scalable multi-processed fuzzing
+pub struct CentralizedLlmpHook<I, SP>
+where
+    I: Input,
+    SP: ShMemProvider + 'static,
+{
+    #[cfg(feature = "llmp_compression")]
+    compressor: GzipCompressor,
+    phantom: PhantomData<(I, SP)>,
+}
+
+impl<I, SP> LlmpHook for CentralizedLlmpHook<I, SP>
+where
+    I: Input,
+    SP: ShMemProvider + 'static,
+{
+    fn on_new_message(
+        &mut self,
+        client_id: ClientId,
+        msg_tag: &mut Tag,
+        msg_flags: &mut Flags,
+        msg: &mut [u8],
+    ) -> Result<LlmpMsgHookResult, Error> {
+        if *msg_tag == _LLMP_TAG_TO_MAIN {
+            #[cfg(feature = "llmp_compression")]
+            let compressor = &self.compressor;
+            #[cfg(not(feature = "llmp_compression"))]
+            let event_bytes = msg;
+            #[cfg(feature = "llmp_compression")]
+            let compressed;
+            #[cfg(feature = "llmp_compression")]
+            let event_bytes = if *msg_flags & LLMP_FLAG_COMPRESSED == LLMP_FLAG_COMPRESSED {
+                compressed = compressor.decompress(msg)?;
+                &compressed
+            } else {
+                &*msg
+            };
+            let event: Event<I> = postcard::from_bytes(event_bytes)?;
+            match Self::handle_in_broker(client_id, &event)? {
+                BrokerEventResult::Forward => Ok(LlmpMsgHookResult::ForwardToClients),
+                BrokerEventResult::Handled => Ok(LlmpMsgHookResult::Handled),
+            }
+        } else {
+            Ok(LlmpMsgHookResult::ForwardToClients)
+        }
+    }
+}
+
+impl<I, SP> Debug for CentralizedLlmpHook<I, SP>
+where
```

Comment:
Here, too.
you don't need sp to be shmem provider or i to be input to implement Debug